File: backend/src/services/favoriteService.js
// src/services/favoriteService.js
// Favoritos (coração) persistidos por utilizador.
const prisma = require('../config/database');
const { NotFoundError } = require('../utils/errors');

async function listIds(userId) {
  const favs = await prisma.favorite.findMany({ where: { userId }, select: { productId: true } });
  return favs.map((f) => f.productId);
}

async function add(userId, productId) {
  const product = await prisma.product.findUnique({ where: { id: productId } });
  if (!product) throw new NotFoundError('Produto');
  await prisma.favorite.upsert({
    where: { userId_productId: { userId, productId } },
    update: {},
    create: { userId, productId },
  });
  return { productId, favorite: true };
}

async function remove(userId, productId) {
  await prisma.favorite.deleteMany({ where: { userId, productId } });
  return { productId, favorite: false };
}

module.exports = { listIds, add, remove };

File: backend-java/src/main/java/ao/kixima/marketplace/FavoriteService.java
package ao.kixima.marketplace;

import ao.kixima.catalog.ProductRepository;
import ao.kixima.common.error.NotFoundException;
import ao.kixima.marketplace.dto.FavoriteResultDto;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.List;
import java.util.UUID;

/** Espelha backend/src/services/favoriteService.js — favoritos (coração) persistidos por utilizador. */
@Service
public class FavoriteService {

    private final FavoriteRepository favoriteRepository;
    private final ProductRepository productRepository;

    public FavoriteService(FavoriteRepository favoriteRepository, ProductRepository productRepository) {
        this.favoriteRepository = favoriteRepository;
        this.productRepository = productRepository;
    }

    @Transactional(readOnly = true)
    public List<String> listIds(String userId) {
        return favoriteRepository.findByUserId(userId).stream().map(Favorite::getProductId).toList();
    }

    @Transactional
    public FavoriteResultDto add(String userId, String productId) {
        if (!productRepository.existsById(productId)) throw new NotFoundException("Produto");
        if (favoriteRepository.findByUserIdAndProductId(userId, productId).isEmpty()) {
            favoriteRepository.save(new Favorite(UUID.randomUUID().toString(), userId, productId, Instant.now()));
        }
        return new FavoriteResultDto(productId, true);
    }

    @Transactional
    public FavoriteResultDto remove(String userId, String productId) {
        favoriteRepository.deleteByUserIdAndProductId(userId, productId);
        return new FavoriteResultDto(productId, false);
    }
}
